eventlog/eventlog.py

```diff
@@ -0,0 +1,759 @@
+# * Each Event type has a constructor, which parses the input
+#   buffer. The constructor is used when there is no doubt about
+#   what should be parsed and how it should be interpreted.
+# * However, the main parser entry point is the Parse class method.
+#   This method peeks into the input buffer and makes decisions
+#   about what object the buffer should be parsed into, then invokes
+#   the appropriate constructor for that type.
+#   Examples of decision points include checking EFI variable names,
+#   e.g. "BootOrder" is handled differently from "Boot0001".
+# * Each Event class has a "validate" method used in
+#   testing its internal consistency (not really available for all
+#   event types, so in those cases testing returns "true")
+# * Each Event class has a "toJson" method to convert it into
+#   data structures accepted by the JSON pickler
+#   (i.e. dictionaries, lists, strings)
+
+# NOTE on JSON output of raw events. tpm2_eventlog limits raw event
+# output to 1024 characters when there is no intelligent processing
+# done on it.  We follow suit for reasons of compatibility -- and
+# because no one in their right mind will process a 6kB raw string in
+# JSON/hex.
+
+class GenericEvent:
+    def __init__ (self, eventheader: Tuple[int, int, dict, int, int], buffer: bytes, idx: int):
+        self.evpcr   = eventheader[1]
+        self.digests = eventheader[2]
+        self.evsize  = eventheader[3]
+        self.evidx   = eventheader[4]
+        assert len(buffer) >= idx + self.evsize, f'Event log truncated, GenericEvent, evt.idx = {self.evidx}'
+        self.evbuf   = buffer[idx:idx+self.evsize]
+
+        try:
+            self.evtype = Event(eventheader[0])
+        except Exception as _:
+            self.evtype = Event.EV_UNKNOWN
+
+    @classmethod
+    def Parse(cls, eventheader: Tuple[int, int, dict, int, int], buffer: bytes, idx: int):
+        return cls(eventheader, buffer, idx)
+
+    # pylint: disable=no-self-use
+    def validate (self) -> Tuple[bool,bool,str]:
+        return True,True,''
+
+    def toJson (self):
+        return {
+            'EventType':   self.evtype.name,
+            'EventNum':    self.evidx,
+            'PCRIndex':    self.evpcr,
+            'EventSize':   self.evsize,
+            'DigestCount': len(self.digests),
+            'Digests':     list(map(lambda o: o[1], self.digests.items())),
+            'Event':       self.evbuf[:1024].hex()
+        }
+
+# ########################################
+# Events that can be validated by CONTENT_MATCHES_DIGEST
+# TCG Guidance on Integrity Measurements and Event Log Processing, V1, Rev 0.118, 12/15/2021, Section 7.2.5.1
+# EV_S_CRTM_VERSION
+# EV_EFI_VARIABLE_DRIVER_CONFIG
+# EV_SEPARATOR
+# EV_EFI_GPT_EVENT
+# EV_EFI_VARIABLE_BOOT
+# ########################################
+
+class ValidatedEvent (GenericEvent):
+    def validate(self) -> Tuple[bool,bool,str]:
+        for algid in self.digests.keys():
```

Comment:
`for algid in self.digests:` but you later use it as a key for the same dict, so `for algid, refdigest in self.digests.items():`

Reply:
done in 3 places.